docs/research/blake-256-hash-function.md

```diff
@@ -1,13 +1,302 @@
+
+![Merkle Damgard Construction](/img/Merkle-Damgard Construction.svg)
+
+**<em>Figure 1: Merkle Damgard Construction</em>**
+
+As shown in Figure 3 above, the output of the compression function **_f_** known as a chaining value **_CV_** and the initial inputs are the Initialization Vector (IV) shown as **_CV<sub>0</sub>_**, with **_h(x)_** being the output after all message blocks are processed. M-D maintains the collision resistance of the compression function however it lacks the pre-image and second pre-image resistance. As a result generic attacks such as the multicollision, the long message second-preimage, herding and length extension are possible on M-D based hash functions [^7] [^8] [^9] [^10] [^11]. 
+
+BLAKE's uses HAIFA, which maintains the valuable properties of the M-D construction and adds to the security and scalability of the transformation. HAIFA is essentially M-D construction but with a mandatory counter (number of bits hashed so far) and an optional salt (random data that used as an additional input).
+
+![HAIFA Construction](/img/HAIFA (BLAKE).svg)
+
+**<em>Figure 2: HAIFA Construction</em>**
+
+This iteration mode solves many of the internal collision problems with M-D construction and provides resistance to herding, long message second pre-image, and length extension attacks as the additional salt and counter simulate distinct compression functions for each data block processed [^4] [^12]. HAIFA is both prefix- and suffix-free and, as a result, is collision-resistant as well as indifferentiable from a random oracle, assuming the underlying compression function is ideal [^1] [^13] [^14]. 
+
+#### **The local wide-pipe compression function**
+
+
+![Local wide-pipe](/img/Wide-pipe.svg)
+
+**<em>Figure 3: Local wide-pipe construction of BLAKE's compression function, inherited from LAKE hash function [^6]</em>**
+
+BLAKE uses the LAKE local wide-pipe structure for its strong security guarantees against collision attacks [^1] [^6]. The compression function takes as input four values: a chaining value, a message block, an optional salt and a mandatory counter. In the case of BLAKE-256: 
+
+| chaining value| message block| Salt  | counter
+|---------------|--------------|-------|--------|
+|     256-bit   |   512-bit    |128-bit|64 -bit |
+
+The structure then works in three steps:
+
+1) Initialization of a large internal state represented by 16 words in a 4x4 word matrix.
```

Comment:
Numbered list items in markdown should be prefixed with `1.` on every line, rather than using brackets and incrementing the number manually.  For example
```
1. Item 1
1. Item 2
1. Item 3
```